### PR TITLE
Drop IrrCompileConfig

### LINKED ISSUE #158: Drop IrrCompileConfig in favor of proper CMake configuration
`IrrCompileConfig.h` is a reminiscent of days before CMake. It was necessary for platform detection and for customization. Nowadays, CMake is a way better way for both.

Depends: minetest/minetest#13227

### PR DESCRIPTION
**Pull request by numberZero**
Close #158.
It may be easier to review commit-by-commit.

### CODE REVIEW

**Comment by numberZero** [comment]
Raw logic behind IrrCompileConfig:
platform-setup-raw drawio

**Comment by numberZero** [comment]
My understanding of logic behind IrrCompileConfig:
platform-setup drawio

**Comment by JosiahWI** [comment]
You may have already considered this, but moving some of the platform-dependent configuration to a separate script might be good.

**Comment by numberZero** [comment]
> moving some of the platform-dependent configuration to a separate script 

From CMakeLists? Maybe but I’d prefer to get rid of most of it instead.

And @sfan5, I think this PR is ready to review now.

**Comment by JosiahWI** [comment]
> > moving some of the platform-dependent configuration to a separate script
> 
> From CMakeLists? Maybe but I’d prefer to get rid of most of it instead.

I'm suggesting a CMake script, to keep src/CMakeLists.txt uncluttered.

File: source/Irrlicht/CMakeLists.txt [hunk 1]
```diff
@@ -113,7 +275,7 @@ elseif(APPLE)
 else()
 	# Unix probably
 	find_package(X11 REQUIRED)
-	if(XINPUT2_ENABLED AND NOT X11_Xi_FOUND)
+	if(${USE_XINPUT2} AND NOT X11_Xi_FOUND)
```
Comment (sfan5) [inline source/Irrlicht/CMakeLists.txt:278]:
```suggestion
	if(USE_XINPUT2 AND NOT X11_Xi_FOUND)
```

**Review by sfan5 — COMMENTED** [review]
Code changes look good otherwise, not tested yet

File: source/Irrlicht/CMakeLists.txt [hunk 2]
```diff
@@ -56,46 +63,201 @@ if(NOT REVISION_SANITY_CHECK)
...
+endif()
+
+if(USE_SDL2)
+	set(DEVICE "SDL")
+endif()
+
+option(USE_X11 "Use X11" TRUE)
+if(USE_X11)
+	add_definitions(-D_IRR_COMPILE_WITH_X11_)
+endif()
```
[the remark this answers is not in the document]
Comment (numberZero) [reply source/Irrlicht/CMakeLists.txt:123]: It does, like the IrrCompileConfig did. It is only checked in `CIrrDeviceLinux` actually so I may limit it to that.

File: include/IrrlichtDevice.h [hunk 3]
```diff
@@ -310,37 +310,7 @@ namespace irr
...
-				case video::EDT_WEBGL1:
-#ifdef _IRR_COMPILE_WITH_WEBGL1_
-					return true;
-#else
-					return false;
-#endif
-				default:
-					return false;
-			}
+			return true;
```
[the remark this answers is not in the document]
Comment (numberZero) [reply include/IrrlichtDevice.h:313]: Yes. It’s not of much use currently anyway.

File: include/IVideoDriver.h [hunk 4]
```diff
@@ -56,32 +56,13 @@ namespace video
...
+		//! Texture 0 transformation
+		//! Use ETS_TEXTURE_0 + texture_number to access other texture transformations
 		ETS_TEXTURE_0,
-		//! Texture transformation
+		//! Texture 1 transformation
 		ETS_TEXTURE_1,
-		//! Texture transformation
-		ETS_TEXTURE_2,
-		//! Texture transformation
-		ETS_TEXTURE_3,
```
[the remark this answers is not in the document]
Comment (numberZero) [reply include/IVideoDriver.h:66]: Did you read the new comment at line 60?

File: source/Irrlicht/CMakeLists.txt [hunk 5]
```diff
@@ -1,9 +1,16 @@
...
 # Compiler flags
 
 add_definitions(-DIRRLICHT_EXPORTS)
-if(NOT BUILD_SHARED_LIBS)
+if(BUILD_SHARED_LIBS)
+	if(WIN32)
+		add_definitions("-DIRRLICHT_API=__declspec(dllexport)")
+	elseif(CMAKE_CXX_COMPILER_ID MATCHES "^(GNU|Clang|AppleClang)$")
+		add_definitions("-DIRRLICHT_API=__attribute__ ((visibility(\"default\")))")
+	endif()
```
[the remark this answers is not in the document]
Comment (numberZero) [reply source/Irrlicht/CMakeLists.txt:12]: Style points. Don’t like to write #ifdef WIN32 #ifndef _IRR_STATIC_LIB #ifdef IRRLICHT_EXPORTS.
Although the fact the counterpart is only defined at line 472 doesn’t feel right...

File: source/Irrlicht/CMakeLists.txt [hunk 6]
```diff
@@ -56,46 +63,201 @@ if(NOT REVISION_SANITY_CHECK)
...
+	if(DEVICE MATCHES "^WINDOWS|X11|ANDROID$")
+		add_definitions(-D_IRR_COMPILE_WITH_EGL_MANAGER_ -D_IRR_OGLES1_USE_EXTPOINTER_)
+	elseif(DEVICE STREQUAL "IOS")
+		add_definitions(-D_IRR_COMPILE_WITH_EAGL_MANAGER_)
+	endif()
+endif()
+
+if(ENABLE_GLES2)
+	add_definitions(-D_IRR_COMPILE_WITH_OGLES2_)
+	if(DEVICE MATCHES "^WINDOWS|X11|ANDROID$" OR EMSCRIPTEN)
```
Comment (numberZero) [inline source/Irrlicht/CMakeLists.txt:212]:
```suggestion
	if(DEVICE MATCHES "^(WINDOWS|X11|ANDROID)$" OR EMSCRIPTEN)
```

File: include/IVideoDriver.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (sfan5) [reply include/IVideoDriver.h:66]: I did not.
in that case you forgot to get rid of `ETS_TEXTURE_1`

File: include/IrrlichtDevice.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfan5) [reply include/IrrlichtDevice.h:313]: During https://github.com/minetest/minetest/pull/13280 I realized a way to tell which drivers are compiled into Irrlicht is pretty much required

File: source/Irrlicht/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfan5) [reply source/Irrlicht/CMakeLists.txt:123]: Please do.

File: include/IrrlichtDevice.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (numberZero) [reply include/IrrlichtDevice.h]: Reverted removal, moved to a cpp instead.

File: source/Irrlicht/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (numberZero) [reply source/Irrlicht/CMakeLists.txt]: Limited to `IRROTHEROBJ`.

[the remark this answers is not in the document]
Comment (numberZero) [reply source/Irrlicht/CMakeLists.txt]: e484698

File: source/Irrlicht/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (numberZero) [reply source/Irrlicht/CMakeLists.txt:13]: b279810

**Review by sfan5 — APPROVED** [review]
Tested on linux/gl, linux/gles2, windows (mingw)

**Comment by numberZero** [comment]
@sfan5 “linux/gl, linux/gl”?

**Pull request merged**